src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.cs

```diff
@@ -0,0 +1,333 @@
+            context.RegisterCompilationStartAction(CompilationStartAction);
+        }
+
+        private void CompilationStartAction(CompilationStartAnalysisContext context)
+            => context.RegisterOperationBlockStartAction(OnOperationBlockStart);
+
+        private void OnOperationBlockStart(OperationBlockStartAnalysisContext operationBlockStartAnalysisContext)
+        {
+            var wellKnownTypeProvider = WellKnownTypeProvider.GetOrCreate(operationBlockStartAnalysisContext.Compilation);
+            var linqChainMethods = GetLinqMethods(wellKnownTypeProvider, s_linqChainMethods);
+            var noEnumerationMethods = GetLinqMethods(wellKnownTypeProvider, s_noEnumerationLinqMethods);
+            var enumeratedMethods = GetEnumeratedMethods(wellKnownTypeProvider, s_immutableCollectionsTypeNamesAndConvensionMethods, s_enumeratedParametersLinqMethods, s_contructorsEnumeratedParameterTypes);
+            var noEffectLinqChainMethods = GetLinqMethods(wellKnownTypeProvider, s_noEffectLinqChainMethods);
+            var compilation = operationBlockStartAnalysisContext.Compilation;
+            var additionalDeferTypes = GetTypes(compilation, s_additionalDeferredTypes);
+
+            var operationBlocks = operationBlockStartAnalysisContext.OperationBlocks;
+            if (operationBlocks.IsEmpty)
+            {
+                return;
+            }
+
+            var syntaxTree = operationBlocks[0].Syntax.SyntaxTree;
+            var options = operationBlockStartAnalysisContext.Options;
+            var customizedEnumerationMethods = options.GetEnumerationMethodsOption(
+                    MultipleEnumerableDescriptor,
+                    syntaxTree,
+                    compilation);
+
+            var customizedLinqChainMethods = options.GetLinqChainMethodsOption(
+                    MultipleEnumerableDescriptor,
+                    syntaxTree,
+                    compilation);
+
+
+            var assumeMethodEnumeratesArguments = options.GetBoolOptionValue(
+                EditorConfigOptionNames.AssumeMethodEnumeratesArguments,
+                MultipleEnumerableDescriptor,
+                syntaxTree,
+                compilation,
+                defaultValue: false);
+
+            // In CFG blocks there is no foreach loop related Operation, so use the
+            // the GetEnumerator method to find the foreach loop
+            var getEnumeratorSymbols = GetGetEnumeratorMethods(wellKnownTypeProvider);
+            var wellKnownSymbolsInfo = new WellKnownSymbolsInfo(
+                linqChainMethods,
+                noEnumerationMethods,
+                enumeratedMethods,
+                noEffectLinqChainMethods,
+                additionalDeferTypes,
+                getEnumeratorSymbols,
+                customizedEnumerationMethods,
+                customizedLinqChainMethods,
+                assumeMethodEnumeratesArguments);
```

Comment:
All of this could be created in CompilationStartAction

Reply:
I remember @mavasani comments in some comments that these can't be in CompilationStartAction.
https://github.com/dotnet/roslyn-analyzers/pull/5882#discussion_r808687053
Is that correct?

Reply:
@Cosifne Retrieving the options needs to be in OperationStartBlock action as they can vary for each SyntaxTree. Retrieving the well known symbols should be done in CompilationStartAction.

Reply:
I see, thanks for clarification